{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "# The 'fields' parameter is passed as a comma-separated string of single names.\n",
    "fields = [\n",
    "    \"file_name\",\n",
    "    #\"analysis.workflow_type\",\n",
    "    #\"experimental_strategy\",\n",
    "    \"cases.project.primary_site\",\n",
    "    \"cases.project.disease_type\",\n",
    "    \"cases.diagnoses.primary_diagnosis\",\n",
    "    \"cases.submitter_id\",\n",
    "    \"cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"cases.diagnoses.tumor_stage\",\n",
    "    \"cases.diagnoses.tumor_grade\",\n",
    "    \"cases.diagnoses.progression_or_recurrence\"\n",
    "    \"cases.diagnoses.vital_status\",\n",
    "    \"cases.diagnoses.days_to_birth\",\n",
    "    \"cases.diagnoses.days_to_death\",\n",
    "    \"cases.diagnoses.morphology\",\n",
    "    \"cases.diagnoses.tissue_or_organ_of_origin\",\n",
    "    \"cases.samples.longest_dimension\"\n",
    "    #\"cases.project.project_id\"\n",
    "]\n",
    "\n",
    "fields = ','.join(fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#miRNA filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"miRNA Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "          {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.experimental_strategy\",\n",
    "                \"value\": [\"miRNA-Seq\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.primary_site\",\n",
    "            \"value\": ['Brain', 'Breast', 'Kidney', 'Lung', 'Thyroid', 'Uterus',\n",
    "       'Prostate', 'Ovary', 'Lymph Nodes', 'Soft Tissue', 'Esophagus',\n",
    "       'Stomach', 'Bone Marrow']\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.program.name\",\n",
    "            \"value\": [\"TCGA\"]\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue project filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"Gene Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [\"HTSeq - FPKM\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.project_id\",\n",
    "            \"value\": ['TCGA-COAD']\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all files RNA-seq\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"Gene Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [\"HTSeq - FPKM\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.program.name\",\n",
    "            \"value\": [\"TCGA\"]\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "##remove manifest to have file fields\n",
    "params = {\n",
    "   \"return_type\": \"manifest\",\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The parameters are passed to 'json' rather than 'params' in this case\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "\n",
    "#print(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70362"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"manifest.txt\",\"w\") as manifest:\n",
    "    manifest.write(response.content.decode(\"utf-8\"))\n",
    "    manifest.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "with open(\"files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases.0.diagnoses.0.morphology</th>\n",
       "      <th>cases.0.diagnoses.0.primary_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.tissue_or_organ_of_origin</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_grade</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.project.disease_type</th>\n",
       "      <th>cases.0.project.primary_site</th>\n",
       "      <th>cases.0.samples.0.longest_dimension</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>file_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz</th>\n",
       "      <td>8140/3</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Sigmoid colon</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iia</td>\n",
       "      <td>Colon Adenocarcinoma</td>\n",
       "      <td>Colorectal</td>\n",
       "      <td>0.7</td>\n",
       "      <td>TCGA-CM-6165-01A-11R-1653-07</td>\n",
       "      <td>TCGA-CM-6165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz</th>\n",
       "      <td>8140/3</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Cecum</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>Colon Adenocarcinoma</td>\n",
       "      <td>Colorectal</td>\n",
       "      <td>0.6</td>\n",
       "      <td>TCGA-G4-6294-01A-11R-1774-07</td>\n",
       "      <td>TCGA-G4-6294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz</th>\n",
       "      <td>8140/3</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Cecum</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>Colon Adenocarcinoma</td>\n",
       "      <td>Colorectal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-AA-A022-01A-21R-A16W-07</td>\n",
       "      <td>TCGA-AA-A022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz</th>\n",
       "      <td>8140/3</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Ascending colon</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Colon Adenocarcinoma</td>\n",
       "      <td>Colorectal</td>\n",
       "      <td>0.3</td>\n",
       "      <td>TCGA-AD-6963-01A-11R-1928-07</td>\n",
       "      <td>TCGA-AD-6963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz</th>\n",
       "      <td>8140/3</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Colon, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>Colon Adenocarcinoma</td>\n",
       "      <td>Colorectal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A6-2684-01C-08R-A277-07</td>\n",
       "      <td>TCGA-A6-2684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 cases.0.diagnoses.0.morphology  \\\n",
       "file_name                                                                         \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                         8140/3   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                         8140/3   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                         8140/3   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                         8140/3   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                         8140/3   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.primary_diagnosis  \\\n",
       "file_name                                                                                \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                   Adenocarcinoma, NOS   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                   Adenocarcinoma, NOS   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                   Adenocarcinoma, NOS   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                   Adenocarcinoma, NOS   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                   Adenocarcinoma, NOS   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.tissue_or_organ_of_origin  \\\n",
       "file_name                                                                                        \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                                 Sigmoid colon   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                                         Cecum   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                                         Cecum   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                               Ascending colon   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                                    Colon, NOS   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.tumor_grade  \\\n",
       "file_name                                                                          \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                    not reported   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                    not reported   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                    not reported   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                    not reported   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                    not reported   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.tumor_stage  \\\n",
       "file_name                                                                          \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                       stage iia   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                        stage iv   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                        stage ii   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                    not reported   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                         stage i   \n",
       "\n",
       "                                                 cases.0.project.disease_type  \\\n",
       "file_name                                                                       \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz         Colon Adenocarcinoma   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz         Colon Adenocarcinoma   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz         Colon Adenocarcinoma   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz         Colon Adenocarcinoma   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz         Colon Adenocarcinoma   \n",
       "\n",
       "                                                 cases.0.project.primary_site  \\\n",
       "file_name                                                                       \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                   Colorectal   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                   Colorectal   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                   Colorectal   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                   Colorectal   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                   Colorectal   \n",
       "\n",
       "                                                  cases.0.samples.0.longest_dimension  \\\n",
       "file_name                                                                               \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                                  0.7   \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                                  0.6   \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                                  NaN   \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                                  0.3   \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                                  NaN   \n",
       "\n",
       "                                                 cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "file_name                                                                                                          \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz                       TCGA-CM-6165-01A-11R-1653-07                \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz                       TCGA-G4-6294-01A-11R-1774-07                \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz                       TCGA-AA-A022-01A-21R-A16W-07                \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz                       TCGA-AD-6963-01A-11R-1928-07                \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz                       TCGA-A6-2684-01C-08R-A277-07                \n",
       "\n",
       "                                                 cases.0.submitter_id  \n",
       "file_name                                                              \n",
       "020aa019-a3a4-4055-92ee-be824a597501.FPKM.txt.gz         TCGA-CM-6165  \n",
       "00106523-5b1d-44ad-a9f1-7d84db08722c.FPKM.txt.gz         TCGA-G4-6294  \n",
       "3eb03db7-efc1-4f20-bbc9-61551905eb9b.FPKM.txt.gz         TCGA-AA-A022  \n",
       "ddc5dfb2-e7e4-4159-930e-120c7d2a8640.FPKM.txt.gz         TCGA-AD-6963  \n",
       "29dab10a-a0a0-4dbd-b574-97c22b70a3f1.FPKM.txt.gz         TCGA-A6-2684  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files = pd.read_csv(\"files.txt\", sep='\\t')\n",
    "df_files.drop(\"id\", axis=1, inplace=True)\n",
    "df_files.set_index(\"file_name\", inplace=True)\n",
    "#df_files.columns=['primary_site','tcga_id','disease_type','primary_diagnosis']\n",
    "#df_files=df_files.reindex(columns=['primary_site','disease_type','primary_diagnosis','case_id'], copy=False)\n",
    "df_files.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_files.to_csv(\"files.dat\", header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": json.dumps(filters),\n",
    "    \"fields\": \"primary_site,disease_type,files.cases.demographic.vital_status,diagnoses.vital_status\",\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"10\"\n",
    "    }\n",
    "response = requests.get(\"https://api.gdc.cancer.gov/cases\", headers = {\"Content-Type\": \"application/json\"}, params = params)\n",
    "with open(\"files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))\n",
    "    files.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
