{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "# The 'fields' parameter is passed as a comma-separated string of single names.\n",
    "fields = [\n",
    "    \"file_name\",\n",
    "    #\"analysis.workflow_type\",\n",
    "    #\"experimental_strategy\",\n",
    "    \"cases.project.primary_site\",\n",
    "    \"cases.project.disease_type\",\n",
    "    \"cases.diagnoses.primary_diagnosis\",\n",
    "    \"cases.submitter_id\",\n",
    "    \"cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"cases.diagnoses.tumor_stage\",\n",
    "    \"cases.diagnoses.tumor_grade\",\n",
    "    \"cases.diagnoses.progression_or_recurrence\"\n",
    "    \"cases.diagnoses.vital_status\",\n",
    "    \"cases.diagnoses.days_to_birth\",\n",
    "    \"cases.diagnoses.days_to_death\",\n",
    "    \"cases.diagnoses.morphology\",\n",
    "    \"cases.diagnoses.tissue_or_organ_of_origin\",\n",
    "    \"cases.samples.longest_dimension\"\n",
    "    #\"cases.project.project_id\"\n",
    "]\n",
    "\n",
    "fields = ','.join(fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#miRNA filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"miRNA Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "          {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.experimental_strategy\",\n",
    "                \"value\": [\"miRNA-Seq\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.project_id\",\n",
    "            \"value\": ['TCGA-BRCA']\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.program.name\",\n",
    "            \"value\": [\"TCGA\"]\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue project filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"Gene Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [\"HTSeq - FPKM\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.project_id\",\n",
    "            \"value\": [\"TCGA-BRCA\"]\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "##remove manifest to have file fields\n",
    "params = {\n",
    "   \"return_type\": \"manifest\",\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The parameters are passed to 'json' rather than 'params' in this case\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "\n",
    "#print(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "190733"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"manifest.txt\",\"w\") as manifest:\n",
    "    manifest.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "with open(\"files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_files = pd.read_csv(\"files.txt\", sep='\\t')\n",
    "df_files.drop(\"id\", axis=1, inplace=True)\n",
    "df_files.set_index(\"file_name\", inplace=True)\n",
    "#df_files.columns=['primary_site','tcga_id','disease_type','primary_diagnosis']\n",
    "#df_files=df_files.reindex(columns=['primary_site','disease_type','primary_diagnosis','case_id'], copy=False)\n",
    "df_files.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_files.to_csv(\"files_fpkm.dat\", header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": json.dumps(filters),\n",
    "    \"fields\": \"primary_site,cases.project.project_id,file_name,cases.demographic.vital_status,cases.demographic.gender,cases.diagnoses.age_at_diagnosis,cases.diagnoses.days_to_last_follow_up,cases.demographic.days_to_death,cases.demographic.days_to_birth,cases.submitter_id,samples.portions.analytes.aliquots.submitter_id,cases.diagnoses.last_known_disease_status,cases.diagnoses.tumor_stage,cases.exposures.years_smoked,cases.exposures.cigarettes_per_day,cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"50000\"\n",
    "    }\n",
    "response = requests.get(\"https://api.gdc.cancer.gov/files\", headers = {\"Content-Type\": \"application/json\"}, params = params)\n",
    "#print(response.content.decode(\"UTF-8\"))\n",
    "with open(\"files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>cases.0.diagnoses.0.morphology</th>\n",
       "      <th>cases.0.diagnoses.0.primary_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.tissue_or_organ_of_origin</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_grade</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.project.disease_type</th>\n",
       "      <th>cases.0.project.primary_site</th>\n",
       "      <th>cases.0.samples.0.longest_dimension</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44caf0b5-d05f-49fd-b8ec-c32d0003c5f4.FPKM.txt.gz</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating duct carcinoma, NOS</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iia</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A2-A1FZ-01A-51R-A14D-07</td>\n",
       "      <td>TCGA-A2-A1FZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>de704076-e915-4749-9729-e6ee06a0d359.FPKM.txt.gz</td>\n",
       "      <td>8520/3</td>\n",
       "      <td>Lobular carcinoma, NOS</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iiia</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LQ-A4E4-01A-11R-A266-07</td>\n",
       "      <td>TCGA-LQ-A4E4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          file_name  \\\n",
       "0  44caf0b5-d05f-49fd-b8ec-c32d0003c5f4.FPKM.txt.gz   \n",
       "1  de704076-e915-4749-9729-e6ee06a0d359.FPKM.txt.gz   \n",
       "\n",
       "  cases.0.diagnoses.0.morphology cases.0.diagnoses.0.primary_diagnosis  \\\n",
       "0                         8500/3      Infiltrating duct carcinoma, NOS   \n",
       "1                         8520/3                Lobular carcinoma, NOS   \n",
       "\n",
       "  cases.0.diagnoses.0.tissue_or_organ_of_origin  \\\n",
       "0                                   Breast, NOS   \n",
       "1                                   Breast, NOS   \n",
       "\n",
       "  cases.0.diagnoses.0.tumor_grade cases.0.diagnoses.0.tumor_stage  \\\n",
       "0                    not reported                       stage iia   \n",
       "1                    not reported                      stage iiia   \n",
       "\n",
       "  cases.0.project.disease_type cases.0.project.primary_site  \\\n",
       "0    Breast Invasive Carcinoma                       Breast   \n",
       "1    Breast Invasive Carcinoma                       Breast   \n",
       "\n",
       "   cases.0.samples.0.longest_dimension  \\\n",
       "0                                  NaN   \n",
       "1                                  NaN   \n",
       "\n",
       "  cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "0                       TCGA-A2-A1FZ-01A-51R-A14D-07                \n",
       "1                       TCGA-LQ-A4E4-01A-11R-A266-07                \n",
       "\n",
       "  cases.0.submitter_id  \n",
       "0         TCGA-A2-A1FZ  \n",
       "1         TCGA-LQ-A4E4  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_messangers = pd.read_csv(\"files_fpkm.dat\")\n",
    "df_messangers.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>cases.0.diagnoses.0.morphology</th>\n",
       "      <th>cases.0.diagnoses.0.primary_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.tissue_or_organ_of_origin</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_grade</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.project.disease_type</th>\n",
       "      <th>cases.0.project.primary_site</th>\n",
       "      <th>cases.0.samples.0.longest_dimension</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>732ee0df-c0fa-4fa3-aa77-989a11172285.mirbase21...</td>\n",
       "      <td>8480/3</td>\n",
       "      <td>Mucinous adenocarcinoma</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A7-A3IY-01A-21R-A21U-13</td>\n",
       "      <td>TCGA-A7-A3IY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cd88d4aa-99b6-495e-97fb-1a062aed48a2.mirbase21...</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating duct carcinoma, NOS</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A8-A08O-01A-21R-A057-13</td>\n",
       "      <td>TCGA-A8-A08O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           file_name  \\\n",
       "0  732ee0df-c0fa-4fa3-aa77-989a11172285.mirbase21...   \n",
       "1  cd88d4aa-99b6-495e-97fb-1a062aed48a2.mirbase21...   \n",
       "\n",
       "  cases.0.diagnoses.0.morphology cases.0.diagnoses.0.primary_diagnosis  \\\n",
       "0                         8480/3               Mucinous adenocarcinoma   \n",
       "1                         8500/3      Infiltrating duct carcinoma, NOS   \n",
       "\n",
       "  cases.0.diagnoses.0.tissue_or_organ_of_origin  \\\n",
       "0                                   Breast, NOS   \n",
       "1                                   Breast, NOS   \n",
       "\n",
       "  cases.0.diagnoses.0.tumor_grade cases.0.diagnoses.0.tumor_stage  \\\n",
       "0                    not reported                         stage i   \n",
       "1                    not reported                        stage iv   \n",
       "\n",
       "  cases.0.project.disease_type cases.0.project.primary_site  \\\n",
       "0    Breast Invasive Carcinoma                       Breast   \n",
       "1    Breast Invasive Carcinoma                       Breast   \n",
       "\n",
       "   cases.0.samples.0.longest_dimension  \\\n",
       "0                                  NaN   \n",
       "1                                  NaN   \n",
       "\n",
       "  cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "0                       TCGA-A7-A3IY-01A-21R-A21U-13                \n",
       "1                       TCGA-A8-A08O-01A-21R-A057-13                \n",
       "\n",
       "  cases.0.submitter_id  \n",
       "0         TCGA-A7-A3IY  \n",
       "1         TCGA-A8-A08O  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_miRNA = pd.read_csv(\"files_miRNA.dat\")\n",
    "df_miRNA.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_messangers = df_messangers.loc[~df_messangers[\"cases.0.submitter_id\"].duplicated(keep=\"first\"),:].rename(columns={\"file_name\":\"file_fpkm\"})\n",
    "df_miRNA = df_miRNA.loc[~df_miRNA[\"cases.0.submitter_id\"].duplicated(keep=\"first\"),:].rename(columns={\"file_name\":\"file_miRNA\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_miRNA</th>\n",
       "      <th>cases.0.diagnoses.0.morphology</th>\n",
       "      <th>cases.0.diagnoses.0.primary_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.tissue_or_organ_of_origin</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_grade</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.project.disease_type</th>\n",
       "      <th>cases.0.project.primary_site</th>\n",
       "      <th>cases.0.samples.0.longest_dimension</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>file_fpkm</th>\n",
       "      <th>cases.0.diagnoses.0.morphology</th>\n",
       "      <th>cases.0.diagnoses.0.primary_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.tissue_or_organ_of_origin</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_grade</th>\n",
       "      <th>cases.0.diagnoses.0.tumor_stage</th>\n",
       "      <th>cases.0.project.disease_type</th>\n",
       "      <th>cases.0.project.primary_site</th>\n",
       "      <th>cases.0.samples.0.longest_dimension</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A7-A3IY</th>\n",
       "      <td>732ee0df-c0fa-4fa3-aa77-989a11172285.mirbase21...</td>\n",
       "      <td>8480/3</td>\n",
       "      <td>Mucinous adenocarcinoma</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A7-A3IY-01A-21R-A21U-13</td>\n",
       "      <td>2c9671e1-e001-4565-862a-a4845c9bad4f.FPKM.txt.gz</td>\n",
       "      <td>8480/3</td>\n",
       "      <td>Mucinous adenocarcinoma</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A7-A3IY-01A-21R-A21T-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A8-A08O</th>\n",
       "      <td>cd88d4aa-99b6-495e-97fb-1a062aed48a2.mirbase21...</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating duct carcinoma, NOS</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A8-A08O-01A-21R-A057-13</td>\n",
       "      <td>f6170d9d-53fb-478d-b305-ab3fd5c4ecc8.FPKM.txt.gz</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating duct carcinoma, NOS</td>\n",
       "      <td>Breast, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A8-A08O-01A-21R-A056-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             file_miRNA  \\\n",
       "cases.0.submitter_id                                                      \n",
       "TCGA-A7-A3IY          732ee0df-c0fa-4fa3-aa77-989a11172285.mirbase21...   \n",
       "TCGA-A8-A08O          cd88d4aa-99b6-495e-97fb-1a062aed48a2.mirbase21...   \n",
       "\n",
       "                     cases.0.diagnoses.0.morphology  \\\n",
       "cases.0.submitter_id                                  \n",
       "TCGA-A7-A3IY                                 8480/3   \n",
       "TCGA-A8-A08O                                 8500/3   \n",
       "\n",
       "                     cases.0.diagnoses.0.primary_diagnosis  \\\n",
       "cases.0.submitter_id                                         \n",
       "TCGA-A7-A3IY                       Mucinous adenocarcinoma   \n",
       "TCGA-A8-A08O              Infiltrating duct carcinoma, NOS   \n",
       "\n",
       "                     cases.0.diagnoses.0.tissue_or_organ_of_origin  \\\n",
       "cases.0.submitter_id                                                 \n",
       "TCGA-A7-A3IY                                           Breast, NOS   \n",
       "TCGA-A8-A08O                                           Breast, NOS   \n",
       "\n",
       "                     cases.0.diagnoses.0.tumor_grade  \\\n",
       "cases.0.submitter_id                                   \n",
       "TCGA-A7-A3IY                            not reported   \n",
       "TCGA-A8-A08O                            not reported   \n",
       "\n",
       "                     cases.0.diagnoses.0.tumor_stage  \\\n",
       "cases.0.submitter_id                                   \n",
       "TCGA-A7-A3IY                                 stage i   \n",
       "TCGA-A8-A08O                                stage iv   \n",
       "\n",
       "                     cases.0.project.disease_type  \\\n",
       "cases.0.submitter_id                                \n",
       "TCGA-A7-A3IY            Breast Invasive Carcinoma   \n",
       "TCGA-A8-A08O            Breast Invasive Carcinoma   \n",
       "\n",
       "                     cases.0.project.primary_site  \\\n",
       "cases.0.submitter_id                                \n",
       "TCGA-A7-A3IY                               Breast   \n",
       "TCGA-A8-A08O                               Breast   \n",
       "\n",
       "                     cases.0.samples.0.longest_dimension  \\\n",
       "cases.0.submitter_id                                       \n",
       "TCGA-A7-A3IY                                         NaN   \n",
       "TCGA-A8-A08O                                         NaN   \n",
       "\n",
       "                     cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "cases.0.submitter_id                                                                   \n",
       "TCGA-A7-A3IY                               TCGA-A7-A3IY-01A-21R-A21U-13                \n",
       "TCGA-A8-A08O                               TCGA-A8-A08O-01A-21R-A057-13                \n",
       "\n",
       "                                                             file_fpkm  \\\n",
       "cases.0.submitter_id                                                     \n",
       "TCGA-A7-A3IY          2c9671e1-e001-4565-862a-a4845c9bad4f.FPKM.txt.gz   \n",
       "TCGA-A8-A08O          f6170d9d-53fb-478d-b305-ab3fd5c4ecc8.FPKM.txt.gz   \n",
       "\n",
       "                     cases.0.diagnoses.0.morphology  \\\n",
       "cases.0.submitter_id                                  \n",
       "TCGA-A7-A3IY                                 8480/3   \n",
       "TCGA-A8-A08O                                 8500/3   \n",
       "\n",
       "                     cases.0.diagnoses.0.primary_diagnosis  \\\n",
       "cases.0.submitter_id                                         \n",
       "TCGA-A7-A3IY                       Mucinous adenocarcinoma   \n",
       "TCGA-A8-A08O              Infiltrating duct carcinoma, NOS   \n",
       "\n",
       "                     cases.0.diagnoses.0.tissue_or_organ_of_origin  \\\n",
       "cases.0.submitter_id                                                 \n",
       "TCGA-A7-A3IY                                           Breast, NOS   \n",
       "TCGA-A8-A08O                                           Breast, NOS   \n",
       "\n",
       "                     cases.0.diagnoses.0.tumor_grade  \\\n",
       "cases.0.submitter_id                                   \n",
       "TCGA-A7-A3IY                            not reported   \n",
       "TCGA-A8-A08O                            not reported   \n",
       "\n",
       "                     cases.0.diagnoses.0.tumor_stage  \\\n",
       "cases.0.submitter_id                                   \n",
       "TCGA-A7-A3IY                                 stage i   \n",
       "TCGA-A8-A08O                                stage iv   \n",
       "\n",
       "                     cases.0.project.disease_type  \\\n",
       "cases.0.submitter_id                                \n",
       "TCGA-A7-A3IY            Breast Invasive Carcinoma   \n",
       "TCGA-A8-A08O            Breast Invasive Carcinoma   \n",
       "\n",
       "                     cases.0.project.primary_site  \\\n",
       "cases.0.submitter_id                                \n",
       "TCGA-A7-A3IY                               Breast   \n",
       "TCGA-A8-A08O                               Breast   \n",
       "\n",
       "                     cases.0.samples.0.longest_dimension  \\\n",
       "cases.0.submitter_id                                       \n",
       "TCGA-A7-A3IY                                         NaN   \n",
       "TCGA-A8-A08O                                         NaN   \n",
       "\n",
       "                     cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \n",
       "cases.0.submitter_id                                                                  \n",
       "TCGA-A7-A3IY                               TCGA-A7-A3IY-01A-21R-A21T-07               \n",
       "TCGA-A8-A08O                               TCGA-A8-A08O-01A-21R-A056-07               "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files = df_miRNA.set_index(\"cases.0.submitter_id\").transpose().append(df_messangers.set_index(\"cases.0.submitter_id\").transpose()).transpose()\n",
    "df_files.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_files.to_csv(\"files_manifest.dat\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data_miRNA/002f73df-6c1d-4187-bab8-09908ed64a75/09a8fbb6-a236-4c64-be34-932db3fd3f06.mirbase21.mirnas.quantification.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-c19a8bed167a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"data_miRNA/002f73df-6c1d-4187-bab8-09908ed64a75/09a8fbb6-a236-4c64-be34-932db3fd3f06.mirbase21.mirnas.quantification.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"miRNA_ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"miRNA.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    812\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    813\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 814\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    815\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    816\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1043\u001b[0m             )\n\u001b[1;32m   1044\u001b[0m         \u001b[0;31m# error: Too many arguments for \"ParserBase\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1045\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[call-arg]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1046\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1047\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_failover_to_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1860\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1861\u001b[0m         \u001b[0;31m# open handles\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1862\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open_handles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1863\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1864\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"storage_options\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"encoding\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"memory_map\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"compression\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_open_handles\u001b[0;34m(self, src, kwds)\u001b[0m\n\u001b[1;32m   1355\u001b[0m         \u001b[0mLet\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mreaders\u001b[0m \u001b[0mopen\u001b[0m \u001b[0mIOHanldes\u001b[0m \u001b[0mafter\u001b[0m \u001b[0mthey\u001b[0m \u001b[0mare\u001b[0m \u001b[0mdone\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mtheir\u001b[0m \u001b[0mpotential\u001b[0m \u001b[0mraises\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1356\u001b[0m         \"\"\"\n\u001b[0;32m-> 1357\u001b[0;31m         self.handles = get_handle(\n\u001b[0m\u001b[1;32m   1358\u001b[0m             \u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1359\u001b[0m             \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    640\u001b[0m                 \u001b[0merrors\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"replace\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    641\u001b[0m             \u001b[0;31m# Encoding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 642\u001b[0;31m             handle = open(\n\u001b[0m\u001b[1;32m    643\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    644\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data_miRNA/002f73df-6c1d-4187-bab8-09908ed64a75/09a8fbb6-a236-4c64-be34-932db3fd3f06.mirbase21.mirnas.quantification.txt'"
     ]
    }
   ],
   "source": [
    "pd.read_csv(\"data_miRNA/002f73df-6c1d-4187-bab8-09908ed64a75/09a8fbb6-a236-4c64-be34-932db3fd3f06.mirbase21.mirnas.quantification.txt\", sep=\"\\t\")[\"miRNA_ID\"].to_csv(\"miRNA.txt\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
